File: rust/crates/orca-dispatch/src/modules/mobile_relay_pairing_offer.rs
//! Parity dispatch for `orca_relay::pairing_offer_schema` vs
//! `src/shared/mobile-relay-pairing-offer.ts`.
//!
//! This module exists because the twin's validation is a SIBLING of the module
//! that exports it: `pairing.ts` re-exports `PairingOfferSchema` and owns no rule
//! of its own, so a corpus keyed on `pairing.test.ts` never sees the relay v1
//! sub-object, the size caps or the cross-field rules at all.
//!
//! Every function here takes `nowMs` from the vector. The twin's own
//! `createPairingOfferSchema(now)` takes the same injected clock, so the invite
//! window — the one wall-clock rule in the schema — is compared at exact
//! boundaries rather than at whatever instant the two processes happened to run.

use super::pairing::offer_from_input;
use orca_relay::{
    decode_pairing_offer, encode_pairing_offer, pairing_offer_to_json, parse_pairing_code,
    validate_pairing_offer,
};
use serde_json::{json, Value};

pub fn dispatch(function: &str, input: &Value) -> Value {
    let now_ms = input.get("nowMs").and_then(Value::as_f64).unwrap_or(f64::NAN);
    let field = |key: &str| input.get(key).and_then(Value::as_str).unwrap_or_default().to_string();
    match function {
        "validatePairingOffer" => {
            let payload = input.get("payload").cloned().unwrap_or(Value::Null);
            match validate_pairing_offer(&payload, now_ms) {
                Some(offer) => pairing_offer_to_json(&offer),
                None => Value::Null,
            }
        }
        "encodePairingOffer" => {
            let offer = offer_from_input(input.get("offer").unwrap_or(&Value::Null));
            match encode_pairing_offer(&offer, now_ms) {
                Ok(url) => Value::String(url),
                Err(_) => Value::Null,
            }
        }
        "decodePairingOffer" => match decode_pairing_offer(&field("url"), now_ms) {
            Ok(offer) => pairing_offer_to_json(&offer),
            Err(_) => Value::Null,
        },
        "parsePairingCode" => match parse_pairing_code(&field("input"), now_ms) {
            Some(offer) => pairing_offer_to_json(&offer),
            None => Value::Null,
        },
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}

File: tools/parity/dispatch/mobile-relay-pairing-offer.ts
// TS dispatch for the mobile-relay-pairing-offer parity module: drives the LIVE
// `src/shared/mobile-relay-pairing-offer.ts` schema — the validation that
// `src/shared/pairing.ts` re-exports and owns none of.
//
// THE CLOCK. The schema's invite window is evaluated against `now()`, so it is
// not a pure function of its input and cannot be a parity vector without one.
// `createPairingOfferSchema(now)` is the twin's OWN injection point, so
// `validatePairingOffer` passes the vector's `nowMs` straight into it — no
// patching, no double. The three deep-link arms used to pin the global clock,
// because the twin's exports were bound to the default `() => Date.now()`; they
// now go through `src/shared/pairing-deep-link.ts`, which takes `nowMs` as an
// argument on every entry point, so the vector's instant is simply passed. Both
// routes compare the two halves AT A NAMED INSTANT; what neither can promise is
// that production's two callers observe the same instant — that is why the Rust
// side takes `now_ms` as an argument instead of reading a clock of its own.
//
// The three deep-link arms drive the SHIM with the seam unbound (this config
// installs no setup file), so they compare the pre-ready fallback against Rust.
// The fallback is where this module's own residuals are mirrored, which is why
// the two `xn--` vectors below still carry `allowDivergence`: the case that
// diverges is `validatePairingOffer`, which is the raw schema and does not go
// through the shim.

import {
  decodePairingOffer,
  encodePairingOffer,
  parsePairingCode,
  type PairingOffer
} from '../../../src/shared/pairing-deep-link'
import { createPairingOfferSchema } from '../../../src/shared/mobile-relay-pairing-offer'

type ClockInput = {
  nowMs?: number
  payload?: unknown
  offer?: unknown
  url?: string
  input?: string
}

function requireClock(args: ClockInput): number {
  // Why: an omitted nowMs would silently fall back to the wall clock on the TS
  // side while Rust read NaN — a vector that cannot be compared must not run.
  if (typeof args.nowMs !== 'number') {
    throw new Error('mobile-relay-pairing-offer vectors must pass an explicit `nowMs`')
  }
  return args.nowMs
}

function orNull<T>(run: () => T): T | null {
  try {
    return run()
  } catch {
    return null
  }
}

export function dispatch(fn: string, input: unknown): unknown {
  const args = (input ?? {}) as ClockInput
  const nowMs = requireClock(args)
  switch (fn) {
    case 'validatePairingOffer': {
      const result = createPairingOfferSchema(() => nowMs).safeParse(args.payload)
      return result.success ? result.data : null
    }
    case 'encodePairingOffer':
      return orNull(() => encodePairingOffer(args.offer as PairingOffer, nowMs))
    case 'decodePairingOffer':
      return orNull(() => decodePairingOffer(args.url as string, nowMs))
    case 'parsePairingCode':
      return parsePairingCode(args.input as string, nowMs)
    default:
      throw new Error(`unknown function ${fn}`)
  }
}
